src/OAuth2/Storage/IbmDb2.php

```diff
@@ -0,0 +1,610 @@
+    AccessTokenInterface,
+    ClientCredentialsInterface,
+    UserCredentialsInterface,
+    RefreshTokenInterface,
+    JwtBearerInterface,
+    ScopeInterface,
+    PublicKeyInterface,
+    UserClaimsInterface,
+    OpenIDAuthorizationCodeInterface
+{
+    protected $db;
+    protected $config;
+
+    public function __construct($connection, $config = array())
+    {
+        if (!is_resource($connection)) {
+            // Note: Unlike PDO, IbmDb2 (ibm_db2 extension) cannot be configured via dsn string.
+
+            if (!is_array($connection)) {
+                throw new \InvalidArgumentException('First argument to OAuth2\Storage\IbmDb2 must be a resource or a configuration array');
+            }
+
+            if (!extension_loaded('ibm_db2')) {
+                throw new \RuntimeException('The ibm_db2 extension is required by OAuth2\Storage\IbmDb2.');
+	        }
+
+            // merge optional parameters. Set empty defaults if not present in $connection array.
+            $connection = array_merge(array(
+                'database' =>     '',
+                'username' =>   '',
+                'password' =>   '',
+                'persistent' => false,
+                'driver_options' =>    array(),
+            ), $connection);
+
+            // use persistent or not
+            $isPersistent = $connection['persistent'];
+            $connectFunction = ((bool) $isPersistent) ? 'db2_pconnect' : 'db2_connect';
+
+            // try to connect
+            $connection = $connectFunction($connection['database'], $connection['username'], $connection['password'], $connection['driver_options']);
+
+            // this is how ZF2 handles connection errors
+            if ($connection === false) {
+                throw new \RuntimeException(sprintf(
+                    '%s: Unable to connect to database',
+                    __METHOD__
+                ));
+            }
+
+        }
+
+        $this->db = $connection;
+
+        $this->config = array_merge(array(
+            'client_table' => 'oauth_clients',
+            'access_token_table' => 'oauth_access_tokens',
+            'refresh_token_table' => 'oauth_refresh_tokens',
+            'code_table' => 'oauth_authorization_codes',
+            'user_table' => 'oauth_users',
+            'jwt_table'  => 'oauth_jwt',
+            'jti_table'  => 'oauth_jti',
+            'scope_table'  => 'oauth_scopes',
+            'public_key_table'  => 'oauth_public_keys',
+        ), $config);
+    }
+
+    /* OAuth2\Storage\ClientCredentialsInterface */
+    public function checkClientCredentials($client_id, $client_secret = null)
+    {
+        $stmt = db2_prepare($this->db, sprintf('SELECT * from %s where client_id = ?', $this->config['client_table']));
+        $successfulExecute = db2_execute($stmt, compact('client_id'));
+        $result = db2_fetch_assoc($stmt);
+
+        // make this extensible
+        return $result && $result['client_secret'] == $client_secret;
```

Comment:
Is the purpose of the expression to test if $result exists, and if so, perform the == comparison?  The == comparison operator has higher precedence than && logical operator, meaning if I have the intent correct, the expression should actually be $result && ($result['client_secret'] == $client_secret).  In this way, PHP's short-circuit boolean evaluation can be applied without undefined variable notice. Or have I missed something obvious (it wouldn't be the first time)?